packages/wrapper/src/editorApp.ts

```diff
@@ -249,18 +253,30 @@ export class EditorApp {
                     original: textModelOriginal,
                     modified: textModelModified
                 };
-                if (this.diffEditor.getModel()?.modified !== textModelModified) {
-                    this.diffEditor.getModel()?.modified.dispose();
+                const modifiedUri = this.diffEditor.getModel()?.modified.uri.toString();
+                const originalUri = this.diffEditor.getModel()?.original.uri.toString();
+                if (modifiedUri !== textModelModified.uri.toString()) {
+                    if (modifiedUri !== undefined) {
+                        diposeOldModels.push(modifiedUri);
+                    }
                 }
-                if (this.diffEditor.getModel()?.original !== textModelOriginal) {
-                    this.diffEditor.getModel()?.original.dispose();
+                if (originalUri !== textModelOriginal.uri.toString()) {
+                    if (originalUri !== undefined) {
+                        diposeOldModels.push(originalUri);
+                    }
                 }
                 this.diffEditor.setModel(textModels);
                 this.modelUpdateCallback?.(textModels);
             } else {
                 throw new Error('You cannot update models, because original model ref is not contained, but required for DiffEditor.');
             }
         }
+
+        monaco.editor.getModels().forEach(model => {
+            if (diposeOldModels.includes(model.uri.toString())) {
+                model.dispose();
+            }
+        });
```

Comment:
nit: instead of looping in each model, you can loop on your diposeOldModels and get the model by passing the url to `getModels`

btw, shouldn't you be using model ref instead and just disposing the ref?

Reply:
I have updated the code ⬇️

Reply:
Btw, is it save to compare the uris or should I compare the string representations to check if they are different. WDYT?

Reply:
> What do you mean? if there is no ref left, the model is disposed

Interestingly they are kept globally after disposing the reference, therefore I have to call https://github.com/TypeFox/monaco-languageclient/blob/244051d9ef8344d0df9fd4a72c46830f95cf7132/packages/wrapper/src/editorApp.ts#L275-L277 otherwise the `didClose` notification is not sent from client to server.

> Comparing uri objects using == is not safe. Why do you need to do that?

Fixed the comparison. I only want to update the editor if the uris really changed.

Reply:
@CGNonofr does that ⬆️ sound like something is not working correctly inside monaco?

Reply:
> It doesn't look safe to force dispose it I've seen some part of the code creating a reference and disposing it in a timeout

That is bad news, because I don't know how it handle it then. Btw, this works properly with the views service. This problem only occurs with the wrapper/editor service.

Reply:
It's no that hard to know btw, you can add some debug logs in the createModelReference function to see which ref is not released

Reply:
to reproduce:
https://github.com/cdietrich/monaco-editor-react-new/commit/d20817c2ffc9dc450f3c6926d1dd4e2d74e540ac
and comment out
<img width="360" alt="grafik" src="https://github.com/user-attachments/assets/85d629df-a71d-4ebe-abb8-11c82df9d931" />

Reply:
hmm i see createModelReference called once only

Reply:
@CGNonofr and @cdietrich the problem is gone. It seems that it was a combination of disposing the model reference at the wrong time and always setting the textEditorModel even if not needed. The test workarounds are also no longer required.

Reply:
`monaco.editor.getModel(uri)?.dispose` calls are removed.

Reply:
> The test workarounds are also no longer required.

They were a strong indication that I did something wrong, but I did not understand why last Friday. Anyway, enforcing everyone to deal explicitly with uris is a necessary improvement.